### PR TITLE
give up if cannot generate a proper sample

### PR DESCRIPTION
**Pull request by aplavin**
I've encountered hanging in this loop forever with some models, pretty difficult to reproduce. Sometimes the random points just don't fall into the unit cube, eg when ellipses happen to be almost degenerate (then after `maxiter` we see `ncall = 0`). Sometimes likelihoods higher than `logl_star` are not achieved by random points, then `ncall ~ maxiter`.

Looks like a bail-out limit is useful here to prevent hanging of the whole process. The `maxiter` number is chosen arbitrary in this PR for now, advice is welcome!

### CODE REVIEW

**Comment by mileslucas** [comment]
I'm actually going to suggest a slightly more involved PR, which fits in with a change I wanted to make anyways-

I've wanted to rename `Proposals.Uniform` to `Proposals.Rejection`, since I find "rejection sampling" makes more sense than "uniform sampling", especially in the context of "Gibbs slice sampling", "random walk sampling" (`Slice`, `RWalk`, respectively). In addition, I want to avoid confusing `Proposals.Uniform` with `Distributions.Uniform`.

Why don't we
1. add an immutable struct `Proposals.Rejection` like so
```julia
@with_kw struct Rejection <: AbstractProposal
	maxiter = 100000
    @assert maxiter > 0
end
```
2. deprecate `Uniform` (and update references to `Uniform` throughout the codebase)
```julia
@deprecate Uniform() Rejection(maxiter=Inf)
```
3. Use the code you've already written just using this parameterized version of the struct

**Comment by aplavin** [comment]
Added the parameter to `Uniform` struct. Do you have any other suggestion what it's default value should be? I just put 100k, seems generally reasonable to me.

Didn't rename anything: these are totally unrelated changes, aren't they? Also, there is certain prior art on usage of "uniform" in this context anyway.

**Comment by aplavin** [comment]
Don't know why these PRs always fail on Julia 1.3...

**Comment by mileslucas** [comment]
>  what it's default value should be? I just put 100k, seems generally reasonable to me

Good question! 100,000 iterations is *a lot* in context with the static nested sampler. Typically the number of loglike evals to samples produced is 10% efficient, so for 100,000 iterations we don't want it worse than 10% efficiency, which is 10,000 samples. So we're probably over-shooting, but thats fine. 

So 100,000 is fine!

> Didn't rename anything: these are totally unrelated changes, aren't they?

Fair enough, I can implement those myself in a separate PR!

> Don't know why these PRs always fail on Julia 1.3...

testing this library has proven incredibly frustrating. Trying to prove correctness using the logZ estimates has not been as stable as I hoped. I wouldn't worry too much about it unless multiple platforms/versions fail. I'll shoulder that burden.

File: src/proposals/Proposals.jl [hunk 1]
```diff
@@ -45,24 +45,27 @@ unitcheck(us) = all(u -> 0 < u < 1, us)
 
 Propose a new live point by uniformly sampling within the bounding volume.
 """
```
Comment (mileslucas) [inline src/proposals/Proposals.jl:47]: Can you update the docstring, too?
```julia
"""
    Proposals.Uniform(;maxiter=100000)

Propose a new live point by uniformly sampling within the bounding volume.

## Parameters
- `maxiter` is the maximum number of samples that can be rejected before giving up and throwing an error.
"""
```

Comment (aplavin) [reply src/proposals/Proposals.jl:50]: done!

**Comment by aplavin** [comment]
> Typically the number of loglike evals to samples produced is 10% efficient, so for 100,000 iterations we don't want it worse than 10% efficiency, which is 10,000 samples.

Note that `maxiter` in this PR doesn't mean likelihood evaluations. Sometimes, samples are rejected because they don't fall within the unit cube, and the likelihood isn't even evaluated.

**Review by mileslucas — APPROVED** [review]

**Pull request merged**